src/bezpath.rs

```diff
@@ -214,6 +236,110 @@ impl<'a> IntoIterator for &'a BezPath {
     }
 }
 
+impl IntoIterator for BezPath {
+    type Item = PathEl;
+    type IntoIter = std::vec::IntoIter<PathEl>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+/// Proportion of tolerance budget that goes to cubic to quadratic conversion.
+const TO_QUAD_TOL: f64 = 0.1;
+
+/// Flatten the path, invoking the callback repeatedly.
+///
+/// See [`BezPath::flatten`](struct.BezPath.html#method.flatten) for more discussion.
+/// This signature is a bit more general, allowing flattening of `&[PathEl]` slices
+/// and other iterators yielding `PathEl`.
+pub fn flatten(
+    path: impl IntoIterator<Item = PathEl>,
+    tolerance: f64,
+    mut callback: impl FnMut(PathEl),
+) {
```

Comment:
looking at this, my first thought is that it would be more natural to express this as an iterator, which would remove the need for the callback and simplify the cases where the caller just wants to collect the elements into a vec or a new `BezPath`; and the iterator version should still have all the lazy-evaluation advantages that you might get from the callback?

It's possible I'm misunderstanding the goals, here, but this API is a bit unconventional, and I'm wondering if that's necessary?

Reply:
I agree that an iterator would be more natural and more convenient for the caller. But having to write this algorithm in continuation-passing style would be a pain. I'll note that Lyon also provides a [for_each_flattened](https://docs.rs/lyon_geom/0.15.2/lyon_geom/cubic_bezier/struct.CubicBezierSegment.html#method.for_each_flattened) method with a similar callback signature, though it also provides the iterator.

Reply:
As @raphlinus said, lyon has both iterator and callback APIs for path flattening. In practice, while iterators look a bit nicer/rustier, the callback APIs tend to be a faster and easier to implement especially when there is nesting (like cubics into quadratics into line segments). I end up using the callback APIs whenever I can because of the slight performance advantage.
The main reason I keep the iterator version as well is that it can compose with other iterators in places where the callback wouldn't (for example an iterator that flattens an entire path). It's not entirely clear to me whether these few things where I use the flattening iterators are things that I really need, though.

Reply:
okay, thanks @nical.

I do think this API is going to be confusing to someone who isn't expecting it, but maybe that's a vanishingly small number of people? As in: if you're looking for this, you already have a solid idea of what's going on.